python_modules/libraries/dagster-dg/dagster_dg/utils/__init__.py

```diff
@@ -440,64 +459,170 @@ def parse_json_option(context: click.Context, param: click.Option, value: str):
 # ##### TOML MANIPULATION
 # ########################
 
+TomlPath: TypeAlias = tuple[Union[str, int], ...]
+TomlDoc: TypeAlias = Union[tomlkit.TOMLDocument, dict[str, Any]]
+
 
-def get_toml_value(
-    doc: tomlkit.TOMLDocument,
-    path: Iterable[str],
+def get_toml_node(
+    doc: TomlDoc,
+    path: TomlPath,
     expected_type: Union[type[T], tuple[type[T], ...]],
 ) -> T:
     """Given a tomlkit-parsed document/table (`doc`),retrieve the nested value at `path` and ensure
     it is of type `expected_type`. Returns the value if so, or raises a KeyError / TypeError if not.
     """
-    current: Any = doc
-    for key in path:
-        # If current is not a table/dict or doesn't have the key, error out
-        if not isinstance(current, dict) or key not in current:
-            raise KeyError(f"Key '{key}' not found in path: {'.'.join(path)}")
-        current = current[key]
-
-    # Finally, ensure the found value is of the expected type
-    if not isinstance(current, expected_type):
+    value = _gather_toml_nodes(doc, path)[-1]
+    if not isinstance(value, expected_type):
         expected_types = expected_type if isinstance(expected_type, tuple) else (expected_type,)
         type_str = " or ".join(t.__name__ for t in expected_types)
         raise TypeError(
-            f"Expected '{'.'.join(path)}' to be {type_str}, "
-            f"but got {type(current).__name__} instead."
+            f"Expected '{toml_path_to_str(path)}' to be {type_str}, "
+            f"but got {type(value).__name__} instead."
         )
-    return current
+    return value
 
 
-def has_toml_value(doc: tomlkit.TOMLDocument, path: Sequence[str]) -> bool:
+def has_toml_node(doc: TomlDoc, path: TomlPath) -> bool:
     """Given a tomlkit-parsed document/table (`doc`), return whether a value is defined at `path`."""
-    leading_path, key = path[:-1], path[-1]
-    current = doc
-    for leading_key in leading_path:
-        if not isinstance(current, dict) or leading_key not in current:
-            return False
-        current = current[leading_key]
-    return isinstance(current, dict) and key in current
+    try:
+        _gather_toml_nodes(doc, path)
+    except KeyError:
```

Comment:
not a huge fan of KeyError as control flow, is it possible to have a separate _has_toml_nodes function or a raise_on_missing parametr or something

Reply:
added a `raise_on_missing`